airflow/jobs/local_task_job.py

```diff
@@ -78,16 +79,11 @@ def _execute(self):
         def signal_handler(signum, frame):
             """Setting kill signal handler"""
             self.log.error("Received SIGTERM. Terminating subprocesses")
-            self.on_kill()
             self.task_instance.refresh_from_db()
-            if self.task_instance.state not in State.finished:
-                self.task_instance.set_state(State.FAILED)
-            self.task_instance._run_finished_callback(  # pylint: disable=protected-access
-                error="task received sigterm"
-            )
-            raise AirflowException("LocalTaskJob received SIGTERM signal")
+            os.kill(self.task_instance.pid, signum)
```

Comment:
Passed this to TI so we can properly handle the callbacks for sigterm

Reply:
But if `self.on_kill()`

is not called we would be missing some cleanup steps, no?

https://github.com/apache/airflow/blob/8217db8cb4b1ff302c5cf8662477ac00f701e78c/airflow/jobs/local_task_job.py#L172-L174

Reply:
It will now exit by itself and not terminate immediately.

The SigTerm will kill the task and run the `task.on_kill` method then exits through the LocalTaskJob.handle_task_exit and the task_runner will terminate. Right?

Reply:
Do we need L82 then? If so let's add a comment explaining why it is required.

Reply:
Added

Reply:
I dont understand this change.

on_kill called task_runner.terminate which terminates the process and tidies up afterwards.

In the case of run_as_user, task_instance.pid will be a different user and thus _we won't be able to send it the kill signal_.

Reply:
I have updated this. Actually I was misunderstanding what triggers this area of the code. 
This is triggered by sigterm on task runner and not sigterm on task. 
When we get the task runner sigterm, we exit the task and handle failure with callback.  This ensures tasks are retried.